### ISSUE TITLE
Null string values - decision needed

### ISSUE THREAD

**Issue by kpodp0ra** [teableio/teable#1183]
## Here is what I discovered

1. Backend allows you to put empty strings as a text value
2. Frontend changes them to `null` before sending API request
3. Null values and empty strings are looking the same in UI
![](https://github.com/user-attachments/assets/bb59c0fc-f421-4789-9739-70ceb38a50af)
4. **There is a bug in aggregation:** 
https://github.com/teableio/teable/blob/8b3df68efdbe134362288e8f7c88a25a7457b6ea/apps/nestjs-backend/src/features/aggregation/aggregation.service.ts#L199-L200
Aggregation gives the empty string and the `null` string the same hash (see screenshot above where count is 4, not 2)
5. **isEmpty and isNotEmpty checks only for null values**
Resolution: #1181

## Here is what ChatGPT thinks:

It's crucial to allow users to distinguish between empty strings and `NULL` values in databases because they represent different states with real-life implications. Here are key use cases where this distinction matters:

- **Contact Information**: Empty strings mean the user explicitly chose not to provide data, while `NULL` signifies the data hasn't been addressed yet.  
- **Survey Responses**: Empty strings indicate a deliberate choice to leave a question blank, while `NULL` reflects that the question wasn't shown or interacted with.  
- **E-Commerce**: Sellers can indicate no product description is needed (empty string), while `NULL` flags pending updates.  
- **Application Settings**: Empty strings express "no preference," whereas `NULL` highlights unset configurations.  
- **Medical Records**: Empty strings confirm no allergies, while `NULL` signifies unknown allergy status.  

## My opinion

I know that implementing such a distinction may require some planning and work, but for me, it would be really useful if the null values would be shown as a dark dot or something like DataGrip has. This would allow me to distinct which values are empty on purpose and which are not yet filled.

Remember that you can hide fields in views, and these fields stay hidden in creation-modal. If one people in the company create records and others edit them, it is good to distinguish which fields have been populated.

Also, we have NULL and 0 values in number fields, why we don't implement this mechanic in strings?

## I need feedback and decision what path we should go

If we want to keep null values and empty fields unambiguous, I would consider unifying the backend and frontend by changing the empty field to `null` on the server side.

But if we want to keep null values and empty fields separate, as LLM suggests, we should discuss if we want to and how to distinguish them visually in UI. JetBrains DataGrip does it like this:
![](https://github.com/user-attachments/assets/c0f91405-9cc4-48a3-a187-42b8c251b1ba)

We would also need some context-action to set NULL values:
![](https://github.com/user-attachments/assets/7fba4e22-1f2e-4bbf-8d00-32f1cdc37287)

Also, we would need NULL-button here:
![](https://github.com/user-attachments/assets/a3d9649b-386a-4397-a0be-ceed98dff4b3)

And then there is the side issue of how we do with grouping. Should the null and empty string be treated by the aggregation as the same, or should we keep them in separate groups (as it is currently).

**Comment by tea-artist** [comment]
I currently prefer not to save empty strings at the Teable application and API level, and to convert all empty strings to null before saving, In Excel, empty strings are also treated as null, which seems to be more intuitive for non-tech people. (by the way display `<null>` will also confuse non-tech people

**Comment by alexfacciorusso** [comment]
I see both sides, as in I also don't like saving an empty value, and instead maybe prefer putting a checkbox (a boolean) setting if the question got answered. On the other hand, having some data manipulation that might be unwanted and not immediately exposed to the user/developer is also weird.

What I suggest is maybe adding an option in the field like `"Convert empty strings to NULL values"`, which is **disabled** by default, and the user can enable it and get the behaviour that's currently in the app.

**Comment by tea-artist** [comment]
@alexfacciorusso  I think this is a good idea. But, I think the default state should be enabled for `"Convert empty strings to NULL values"`. Users should turn it off only when they clearly need an empty string.

Similarly, if this option is turned off, display a `<null>` in view when the cell is null to distinguish it.

@kpodp0ra What do you think?

**Comment by tea-artist** [comment]
For the grouping situation, distinguishing between null and empty strings should be in line with expectations because we can already use `<null>` to visually distinguish.

**Comment by alexfacciorusso** [comment]
> @alexfacciorusso I think this is a good idea. But, I think the default state should be enabled for `"Convert empty strings to NULL values"`. Users should turn it off only when they clearly need an empty string.
> 
> Similarly, if this option is turned off, display a `<null>` in view when the cell is null to distinguish it.

Sounds fair to me, yes if the field is disabled then the user can clearly see `<null>` in the cell

**Comment by tea-artist** [comment]
<img width="206" alt="image" src="https://github.com/user-attachments/assets/319068bd-5b79-49ad-a83a-346168ae431c" />

TablePlus uses "empty" to represent an empty string. I think this is also a good idea. We can render a light-colored "empty" mark in the cells of empty strings in the table to distinguish empty strings. In this way, completely empty cells can be used to represent null.

**Comment by kpodp0ra** [comment]
@tea-artist what do you think about optional, non default three checkbox states?
![](https://github.com/user-attachments/assets/7ff69c3d-2f31-4b6a-8795-55263f02f3b5)

**Comment by tea-artist** [comment]
Agree, at present, our processing in the application is all set to true or null. However, considering the need to access external data sources in the future, we do have to handle all cases.

> @tea-artist what do you think about optional, non default three checkbox states? ![](https://private-user-images.githubusercontent.com/21959209/399650905-7ff69c3d-2f31-4b6a-8795-55263f02f3b5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY0MDQ3NzQsIm5iZiI6MTczNjQwNDQ3NCwicGF0aCI6Ii8yMTk1OTIwOS8zOTk2NTA5MDUtN2ZmNjljM2QtMmYzMS00YjZhLTg3OTUtNTUyNjNmMDJmM2I1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTA5VDA2MzQzNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg3OTI5NTBhNjcwYmVlZTY3NDRkMzE0ZmRlYmVkYTc0MWUwYzE5OTE2NTI5ZTgwNmUyYjVhMzViNjcyYWEyZWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.X5Ov-IK70RF2KXzOLvled-el4sozsncY7c3LayBMS58)
